### PR TITLE
Add typing for tmt/steps/execute/internal.py

### PR DESCRIPTION
**Pull request by guy9050**
Blocked on https://github.com/teemtee/tmt/pull/1365

### CODE REVIEW

**Comment by guy9050** [comment]
I'm not sure how to handle this:
```
tmt/steps/execute/internal.py:253: error: Need type annotation for "_results" (hint: "_results: List[<type>] = ...")
Found 1 error in 1 file (checked 14 source files)
```

**Comment by happz** [comment]
> I'm not sure how to handle this:
> 
> ```
> tmt/steps/execute/internal.py:253: error: Need type annotation for "_results" (hint: "_results: List[<type>] = ...")
> Found 1 error in 1 file (checked 14 source files)
> ```

It's exactly as the message says: on line 253, `_results` name is created, but mypy cannot infer what type it is. You have to investigate what it's supposed to store, and add an annotation (`self._results: List[...] = []`).

**Comment by guy9050** [comment]
> > I'm not sure how to handle this:
> > ```
> > tmt/steps/execute/internal.py:253: error: Need type annotation for "_results" (hint: "_results: List[<type>] = ...")
> > Found 1 error in 1 file (checked 14 source files)
> > ```
> 
> It's exactly as the message says: on line 253, `_results` name is created, but mypy cannot infer what type it is. You have to investigate what it's supposed to store, and add an annotation (`self._results: List[...] = []`).

Okay, thanks. What confused me was the face that this variable `self._result` was not one of the attributes that the function is even expecting.

File: tmt/steps/execute/internal.py
Comment (happz) [inline tmt/steps/execute/internal.py:316]: Now we know `self._results` is `List[Result]`, therefore there's no need for `Any` in return value annotation, and we can be more specific.

File: tmt/steps/execute/internal.py
Comment (happz) [inline tmt/steps/execute/internal.py:206]: Follow the dots: `check()` calls either `self.check_beakerlib()` or `self.check_shell()`, what do these two methods return? Can we be more specific than `Any`? I believe we can, inspect those two methods.

File: tmt/steps/execute/internal.py
Comment (happz) [inline tmt/steps/execute/internal.py:126]: I'd say we could use `EnvironmentType` from `tmt.utils` here, instead of `Dict[...]`,

File: tmt/steps/execute/internal.py
Comment (happz) [inline tmt/steps/execute/internal.py:62]: Please, add TODO, this `Any` can be fixed later. I can't find the PR in which this was discussed, but we recently find out what the correct type should be.

File: tmt/steps/execute/internal.py
Comment (happz) [inline tmt/steps/execute/internal.py:88]: Follow what code calls `_show_progress` and with what arguments, and you will find out `progress` can have a more strict type than `Any`, because AFAICT, it's `str`.

File: tmt/steps/execute/internal.py
Comment (happz) [inline tmt/steps/execute/internal.py:150]: See my note on `progress` type above.

File: tmt/steps/execute/internal.py
Comment (guy9050) [reply tmt/steps/execute/internal.py:320]: oh cause it has else - return [] (empty list) so I figured that as not list of strings. changed. Thanks

File: tmt/steps/execute/internal.py
Comment (janhavlin) [reply tmt/steps/execute/internal.py:64]: Unresolving this. The PR is #1304 and the method should return `List[ClickOptionDecoratorType]`, so this should be fixed once #1304 is merged.

Comment (guy9050) [reply tmt/steps/execute/internal.py]: That PR has been merged, so I fixed the code accordingly

File: tmt/steps/execute/internal.py
Comment (thrix) [inline tmt/steps/execute/internal.py:142]: @guy9050 any idea why `extra_environment` cannot actually have `EnvironmentType`?

**Review by thrix — CHANGES_REQUESTED** [review]

File: tmt/steps/execute/internal.py
Comment (guy9050) [reply tmt/steps/execute/internal.py]: Good catch!

File: tmt/steps/execute/internal.py
Comment (janhavlin) [inline tmt/steps/execute/internal.py:194]: Parameter `value` should be `Optional[str]` as it is passed to [`verbose()`](https://github.com/teemtee/tmt/blob/main/tmt/utils.py#L453).

Params `shift` and `level` should both be `int`, they are never set to `None`.

Comment (guy9050) [reply tmt/steps/execute/internal.py:194]: ty, changed.

File: tmt/steps/execute/internal.py
Comment (janhavlin) [inline tmt/steps/execute/internal.py:51]: Can we wait until #1327 gets merged? It is ready for review, so it shouldn't take too long. We could get rid of this `# type: ignore` thanks to that patch.

Comment (thrix) [reply tmt/steps/execute/internal.py:51]: +1, it should be merged very soon!

Comment (guy9050) [reply tmt/steps/execute/internal.py:51]: ack. I see that it's merged so I'll rebase and remove this ignore.

Comment (guy9050) [reply tmt/steps/execute/internal.py:51]: So now after I removed that, a bunch of new errors appear. 
```
tmt/steps/execute/internal.py:100: error: Argument 1 to "listify" has incompatible type "StepData"; expected "Union[Tuple[Any, ...], List[Any], str, Dict[Any, Any]]"
tmt/steps/execute/internal.py:148: error: Item "Common" of "Optional[CommonDerivedType]" has no attribute "plan"
tmt/steps/execute/internal.py:148: error: Item "None" of "Optional[CommonDerivedType]" has no attribute "plan"
tmt/steps/execute/internal.py:269: error: Signature of "go" incompatible with supertype "ExecutePlugin"
tmt/steps/execute/internal.py:269: note:      Superclass:
tmt/steps/execute/internal.py:269: note:          def go(self, *args: Any, **kwargs: Any) -> None
tmt/steps/execute/internal.py:269: note:      Subclass:
tmt/steps/execute/internal.py:269: note:          def go(self, guest: Any) -> None
tmt/steps/execute/internal.py:269: error: Signature of "go" incompatible with supertype "Plugin"
tmt/steps/execute/internal.py:269: note:          def go(self, *args: Any, **kwargs: Any) -> None
tmt/steps/execute/internal.py:269: note:          def go(self, guest: Any) -> None
tmt/steps/execute/internal.py:269: error: Signature of "go" incompatible with supertype "Phase"
tmt/steps/execute/internal.py:269: note:          def go(self, *args: Any, **kwargs: Any) -> None
tmt/steps/execute/internal.py:269: note:          def go(self, guest: Any) -> None
tmt/steps/execute/internal.py:342: error: "Step" has no attribute "_framework"
```

What is an acceptable way to resolve `Item "Common" of "Optional[CommonDerivedType]" has no attribute "plan"` ?

Comment (thrix) [reply tmt/steps/execute/internal.py]: @guy9050  ok, this is actually blocked at least on #1365, updating ... let's wait for that to be merged

File: tmt/steps/execute/internal.py
Comment (guy9050) [inline tmt/steps/execute/internal.py:344]: `tmt/steps/execute/internal.py:344: error: "Step" has no attribute "_framework"`
I checked the class and it's parent classes and it really doesn't have `_framework`. The class `Execute` has it though

File: tmt/steps/execute/internal.py
Comment (thrix) [reply tmt/steps/execute/internal.py]: all mentioned MRs are now merged.

File: tmt/steps/execute/internal.py
Comment (thrix) [reply tmt/steps/execute/internal.py:344]: @happz might be this relates to some of the issues you filled?

Comment (happz) [reply tmt/steps/execute/internal.py:344]: Yeah, it's related to #1373 & co., `self.step` here is just `Step` with no extra detail, while to humans, it's obvious it can ever be `Execute.

Please, add a TODO item here, refering to the issue, and in the meantime `cast(tmt.steps.execute.Execute, self.step)._framework` should help.

Comment (guy9050) [reply tmt/steps/execute/internal.py]: Done. Thanks.

**Comment by guy9050** [comment]
There are two issues currently:
```
tmt/steps/execute/internal.py:187: error: Argument 1 to "join" has incompatible type "Optional[str]"; expected "Union[str, PathLike[str]]"
tmt/steps/execute/internal.py:330: error: Missing positional argument "progress" in call to "execute" of "ExecuteInternal"
```

would like to hear others' thoughts on that

**Comment by happz** [comment]
> There are two issues currently:
> 
> ```
> tmt/steps/execute/internal.py:187: error: Argument 1 to "join" has incompatible type "Optional[str]"; expected "Union[str, PathLike[str]]"

No surprise here:

```
os.path.join(self.discover.workdir, test.path.lstrip('/'))
```

and `self.discover.workdir` is truly `Optional[str]` - if we are at the point where `workdir` is no longer `None`, we can add `assert self.discover.workdir is not None` to make mypy happy.

> tmt/steps/execute/internal.py:330: error: Missing positional argument "progress" in call to "execute" of "ExecuteInternal"

Again, no surprise:

```
    def execute(self, test: Test, guest: Guest, progress: str,
                extra_environment: EnvironmentType) -> None:
```

gets called:

```
            self.execute(
                test, guest,
                extra_environment=extra_environment)
```

`execute` does expect 3 positional arguments, test, guest, and "progress", it's called with just two arguments. It seems the `progress` is not used at all by `execute`, maybe it can be dropped from its signature.
> ```

**Comment by guy9050** [comment]
> execute does expect 3 positional arguments, test, guest, and "progress", it's called with just two arguments. It seems the progress is not used at all by execute, maybe it can be dropped from its signature.

yes that's what I wasn't sure of - add "progress" to the call or remove 


> 

I ended up adding "progress" to the function calling, I'm not sure why it wasn't there/how it worked before..

**Comment by guy9050** [comment]
Rebased. I think that this pull request in not blocked anymore. @thrix

File: tmt/steps/execute/internal.py
Comment (janhavlin) [inline tmt/steps/execute/internal.py:307]: This line might not be needed since the signature of the method got changed. The `extra_environment` variable value will always be either passed in the method parameter or set to `{}`.

Comment (guy9050) [reply tmt/steps/execute/internal.py]: thanks, I agree it's not needed anymore.

File: tmt/steps/execute/internal.py
Comment (janhavlin) [inline tmt/steps/execute/internal.py:105]: No need to use `cast()` here, see how this method is implemented in e.g. [`polarion.py`](https://github.com/teemtee/tmt/blob/main/tmt/steps/report/polarion.py#L24).

Comment (thrix) [reply tmt/steps/execute/internal.py:105]: @guy9050 this is the same comment I left you in https://github.com/teemtee/tmt/pull/1302

Comment (guy9050) [reply tmt/steps/execute/internal.py]: Changed. Thanks.

File: tmt/steps/execute/internal.py
Comment (janhavlin) [inline tmt/steps/execute/internal.py:107]: We got rid of `keys` parameter for all `wake()` methods in https://github.com/teemtee/tmt/pull/1439 because it was unused. Also the `TODO` comment can be deleted?

File: tmt/steps/execute/internal.py
Comment (janhavlin) [inline tmt/steps/execute/internal.py:111]: This `cast()` can be deleted, `listify()` should accept `StepData` type.

File: tmt/steps/execute/internal.py
Comment (janhavlin) [inline tmt/steps/execute/internal.py:116]: Is `finish` really optional? I didn't find a place where it is set to `None`.

**Comment by happz** [comment]
Added to 1.18 milestone - seems to be fairly trivial once rebased & reviewed, and aligned with current development priorities.

File: tmt/steps/execute/internal.py
Comment (happz) [inline tmt/steps/execute/internal.py:73]: Not needed, ‘data.script’ is already normalized.

Comment (janhavlin) [reply tmt/steps/execute/internal.py:73]: Well, `data.script` may be, but the problem is passing the first parameter `self.data` which is of type `StepData`, but `listify` expecting `data: Union[Tuple[Any, ...], List[Any], str, Dict[Any, Any]]`, resulting still in:

```
tmt/steps/execute/internal.py:73: error: Argument 1 to "listify" has incompatible type "StepData"; expected "Union[Tuple[Any, ...], List[Any], str, Dict[Any, Any]]"  [arg-type]
```

Comment (happz) [reply tmt/steps/execute/internal.py:73]: Sorry, I mean the whole `listify` call - `self.data.script` should be already normalized, i.e. a list, thanks to `_normalize_script = tmt.utils.NormalizeKeysMixin._normalize_string_list` in step data class.

Comment (janhavlin) [reply tmt/steps/execute/internal.py]: I see, removed.

File: tmt/steps/execute/internal.py
Comment (happz) [inline tmt/steps/execute/internal.py:266]: Never ever use mutable defaults for keyword arguments.

https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

File: tmt/steps/execute/internal.py
Comment (happz) [inline tmt/steps/execute/internal.py:113]: And `extra_environment` might be set to `None`, by `self._run_tests(self, guest, extra_environment=None)`, therefore `Optional[EnvironmentType] = None` & `extra_environment = extra_environment or {}` is needed.

**Review by happz — APPROVED** [review]

File: tmt/steps/execute/internal.py
Comment (psss) [inline tmt/steps/execute/internal.py:71]: Just calling parent's `wake()`, so not really needed?

**Review by thrix — APPROVED** [review]
lgtm

File: tmt/steps/execute/internal.py
Comment (psss) [inline tmt/steps/execute/internal.py:145]: Why is the `progress` parameter added here but never used in the `execute` method?

**Review by psss — COMMENTED** [review]
Looks good, just two questions.

File: tmt/steps/execute/internal.py
Comment (janhavlin) [reply tmt/steps/execute/internal.py:71]: Right, the method is not needed. Removed.

File: tmt/steps/execute/internal.py
Comment (janhavlin) [reply tmt/steps/execute/internal.py]: No idea. Removed. :)

Comment (psss) [reply tmt/steps/execute/internal.py]: Ok :)

**Review by psss — APPROVED** [review]
Thanks for the adjustments, now looks good.

**Pull request merged**
